apps/emqx_dashboard_rbac/src/emqx_dashboard_rbac.erl

```diff
@@ -31,25 +36,44 @@ role(#?ADMIN{role = Role}) ->
 role([]) ->
     ?ROLE_SUPERUSER;
 role(#{role := Role}) ->
+    Role;
+role(Role) when is_binary(Role) ->
     Role.
 
-valid_role(Role) ->
-    case lists:member(Role, role_list()) of
-        true ->
-            ok;
-        _ ->
-            {error, <<"Role does not exist">>}
-    end.
+valid_dashboard_role(Role) ->
+    valid_role(dashboard, Role).
+
+valid_api_role(Role) ->
+    valid_role(api, Role).
+
 %% ===================================================================
 check_rbac(?ROLE_SUPERUSER, _, _) ->
     true;
+check_rbac(?ROLE_API_SUPERUSER, _, _) ->
+    true;
 check_rbac(?ROLE_VIEWER, <<"GET">>, _) ->
     true;
+check_rbac(?ROLE_API_VIEWER, <<"GET">>, _) ->
+    true;
 %% this API is a special case
 check_rbac(?ROLE_VIEWER, <<"POST">>, <<"/logout">>) ->
     true;
+check_rbac(?ROLE_API_PUBLISHER, <<"POST">>, <<"/publish">>) ->
+    true;
+check_rbac(?ROLE_API_PUBLISHER, <<"POST">>, <<"/publish/bulk">>) ->
+    true;
```

Comment:
Q: So, the PUBLISHER role can only access `POST /publish*` and does not have read access to other resources?

Reply:
This feature was required by @zmstone  and has been confirmed with the product team.